### DISCUSSION TITLE
Background job with hibernate-reactive without blocking the eventloop (java - ignore the area/kotlin label)!

### DISCUSSION THREAD

**Discussion by Froidoh** [quarkusio/quarkus#33340] [Q&A]
Similar to #32665 I have a question

We are facing a similar problem.

We have a background job running in a k8s cluster and need to ensure it's only executed once at the same time.

Team decided to go with the proven DB approach instead of redis and use a shedlock in a db.

So whenever the job runs we need to get a lock-entry in the DB.

If there exists a lock that has not yet exceeded the maximum length a lock might live, we return.
If there is a lock that has exceeded it's maximum lifetime we delete it and attempt to create a new one. (race condition, others might try as well)
If there is no lock, we attempt to create a new one (again, race condition, others might try as well)

Once we have a lock, we execute our blocking logic - **going to the filesystem, reading in some files and again (!) doing some db inserts via hibernate-reactive.**

The processing of the files is what is blocking and depending on the file may take quite a few seconds. During that timeframe, when we block the eventloop, the pod gets killed because it's unresponsive.

To make matters worse we rely on quartz and schedule by some config item - so we can't hardcode something like:

    @Scheduled(every = "1m")

We need it to be something like:

```
    @ConfigProperty(name = "cron.ourJob")
    String cronOurJob;

        var identity = "jobName";

        JobDetail job = JobBuilder.newJob(OurJob.class)
                .withIdentity(identity)
                .build();

        Trigger trigger = TriggerBuilder.newTrigger()
                .withIdentity(identity)
                .withSchedule(cronSchedule(cronOurJob))
                .startNow()
                .build();
        quartz.scheduleJob(job, trigger);
```
The problem is that we always faced the

`HR000069: Detected use of the reactive Session from a different Thread than the one which was used to open the reactive Session - this suggests an invalid integration`

 et al exceptions... if we block the event loop or pods die like flies.

We haven't found a solution yet. I am currently migrating to quarkus 3 to see if we can make this work somehow. 
It's interesting how alien reactive programming feels in Java. Never had any problems in any other language (including Scala and Kotlin) but with vert.x, mutiny and quarkus we keep on struggling.

I am open to switch to something else but hibernate-reactive, for the background-job stuff. Actually I am open for any hints how to resolve this.

Thanks in advance for any ideas!

[a comment by an automated account is not shown]

**Reply by Froidoh** [reply]
Thanks quarkus-bot, but I just mentioned kotlin in a sidenote... we use Java

**Comment by geoand** [comment]
I am pretty sure that Quartz allows you to specify the executor to be used. @machi1990 @mkouba do you guys have any examples where we use the EventLoop as the Quartz executor?

**Reply by machi1990** [reply]
@Froidoh is it possible to have a complete example of the scheduled method? I am looking to see if it contains @NonBlocking annotation or it returns a Uni?

**Reply by machi1990** [reply]
> I am pretty sure that Quartz allows you to specify the executor to be used. @machi1990 @mkouba do you guys have any examples where we use the EventLoop as the Quartz executor?

> I am pretty sure that Quartz allows you to specify the executor to be used. @machi1990 @mkouba do you guys have any examples where we use the EventLoop as the Quartz executor?

NoBlocking jobs are integrated out of the box. If someone is using a @scheduled methods, then returning a Uni or annotation the method with @NonBlocking will do it. If jobs are scheduled programmati, then doing something like the code snippet below can do the job.
  
```java

   @Inject
   Scheduler scheduler; 
    
    ....
    
     JobDefinition asyncJob = scheduler.newJob("jobName")
       .setCron(cronOurJob)
      .setAsyncTask(ec -> {
              // do something here and return a Uni 
   });

  asyncJob.schedule();   
```

**Reply by mkouba** [reply]
Well, it seems that the Quartz API is used directly to schedule a job programatically.  So we'll rather need to see the source the  `OurJob` class.

NOTE: In Quarkus 3, there is an [API for programmatic scheduling](https://quarkus.io/guides/scheduler-reference#programmatic_scheduling) that makes it easy to distinguish a task that should be executed on a blocking thread (`Scheduler.JobDefinition.setTask()`) VS the task that should be executed asynchronously (`Scheduler.JobDefinition.setAsyncTask()`).

In any case, you can't run Hibernate Reactive code concurrently.

@Froidoh You can try use the `VertxContextSupport.subscribeAndAwait()` to [execute asynchronous code (e.g. Hibernate Reactive) from a blocking thread](https://quarkus.io/guides/vertx#executing-asynchronous-code-from-a-blocking-thread).

**Reply by machi1990** [reply]
@Froidoh not related to the core issue but I noticed that you are using the quartz extension and I am wondering if you've any requirement to use quartz extension? It seems that the requirement here is to create a clustered schedulers. The quartz extension has this but that's only work with a blocking data source and not a reactive one, so you cannot use that in the same app that's using the reactive data source (combining reactive and blocking data sources in the same app doesn't work). 
In my opinion, you could use the lightweight [scheduler extension](https://quarkus.io/guides/scheduler) for what you are looking to do with this scheduling need.

**Reply by Froidoh** [reply]
> @Froidoh is it possible to have a complete example of the scheduled method? I am looking to see if it contains @nonblocking annotation or it returns a Uni?

Sure thing! Note that we know that this is a wrong approach and want to learn and rectify this situation! 

It's always a least a tiny bit embarrassing to show code where you know you're obviously doing something wrong

Also note that some decisions were forced upon us - we are replacing a legacy application and all the functionality has to be preserved. 
Quarkus is meant to be a bit of a silver bullet of using way less resources and still be able to use java (about everything else is migrating to Golang or Node.js by company policy)

To give a quick overview:

The application takes files uploaded via an reactive-api, creates a db-entity in the process. This one is set to status QUEUED.
We immediately return a success or error message to the uploading client.
(Works like a charm)

A background job is responsible for actually processing the files - in our example it's emails. 
We have to parse them, extract and store some information to some other tables.

The "parsing and extracting information part" is a CPU-bound task - so it should not run on the EventLoop!
(We do know the theory, from other languages and experience, it's just that we struggle with the implementation in Quarkus, Mutiny, Vert.x ;))

I'm trying to prepare a minimal reproducible example code:

Here is the main job definition as it is right now - blocking the eventloop, shredding our pods, degrading the whole service.

Again: We are aware that sprinkling .await().indefinitely() is the wrong approach in general. 
The team had to get into quarkus coming from completely different backgrounds (languages, frameworks).

You can see the insane hope of just sprinkling .emitOn(Infrastructure.getDefaultWorkerPool()) to make a difference.

Again: The main problem is that we should not block the eventloop but do - and during that time, the pod is killed by k8s because it doesn't react to any health-ping

```Java
@ActivateRequestContext
public class ProcessQueuedEmailJob implements Job {

    private static final Logger log = Logger.getLogger(ProcessQueuedEmailJob.class);
    @ConfigProperty(name = "cron.process-queued-emails-max-minutes-block")
    Integer maxTimeBlock;
    @Inject
    EmailRepository repo;

    @Inject
    ShedlockRepository shedlocks;

    @Inject
    EmailService service;

    private static final String KEY = "QUEUED";

    public void execute(JobExecutionContext context) {
        shedlocks.findById(KEY)
                .emitOn(Infrastructure.getDefaultWorkerPool())
                .onItem().ifNotNull().transformToUni(shedlock -> {
                    if (LocalDateTime.now().isAfter(shedlock.getLockedUntil())) {
                        log.warn("Deleting blocking row in Shedlock for: queued");
                        return shedlocks.delete(KEY)
                                .onFailure().recoverWithUni(e -> Uni.createFrom().item(true))
                                .onItem().invoke(() -> log.debug("Blocking row in Shedlock deleted successfully"))
                                .replaceWithNull();
                    } else {
                        return Uni.createFrom().item(shedlock);
                    }
                })
                .onItem().ifNull().continueWith(() -> {
                    LocalDateTime now = LocalDateTime.now();
                    return shedlocks.persist(new Shedlock(KEY, now.plusMinutes(maxTimeBlock), now, context.getJobDetail().getKey().getName()))
                            .onFailure().recoverWithUni(e -> Uni.createFrom().item(null)).await().atMost(Duration.ofMillis(100));
                })
                // We got a shedlock
                .onItem().ifNotNull().transformToUni(shedlock -> {
                    log.info("STARTING Job for processing queued emails");
                    return repo.takeFirstQueuedToProcessing()
                            .emitOn(Infrastructure.getDefaultWorkerPool())
                            .onItem().ifNotNull().transformToUni(queued -> {
                                String path = queued.getStoredFileName();
                                return switch (FileNameUtils.getExtension(path)) {
                                    case "msg":
                                        yield service.archiveMsgFile(path, queued)
                                                .map(EmailDetails::getEmailInfo)
                                                .flatMap(i -> i.updateStatus(EmailStatus.PROCESSED))
                                                .onFailure().recoverWithUni(e -> queued.updateStatus(EmailStatus.ERROR_PROCESSING));
                                    case "eml":
                                        yield service.archiveEmlFile(path, queued)
                                                .map(EmailDetails::getEmailInfo)
                                                .flatMap(i -> i.updateStatus(EmailStatus.PROCESSED))
                                                .onFailure().recoverWithUni(e -> queued.updateStatus(EmailStatus.ERROR_PROCESSING));
                                    default:
                                        String error = String.format("We only support processing of .eml and .msg files! Archived file for email %s has file name of %s.", queued.emailId, archivedFile.storedFilename);
                                        log.error(error);
                                        yield queued.updateStatus(EmailStatus.ERROR_PROCESSING);
                                };
                            })
                            .invoke(i -> {
                                if (i != null) {
                                    log.debug(String.format("Status of %s is now %s", i.getEmailId(), i.getStatus()));
                                }
                            })
                            .emitOn(Infrastructure.getDefaultWorkerPool())
                            .onTermination().invoke(() -> shedlocks.delete(KEY)
                                    .emitOn(Infrastructure.getDefaultWorkerPool()).await().atMost(Duration.ofMillis(1000)))
                            .emitOn(Infrastructure.getDefaultWorkerPool())
                            .onItem().ignore().andContinueWithNull()
                            ;
                }).await().indefinitely();
    }
}
```

```Java
@ApplicationScoped
public class EmailRepository {
    @WithTransaction
    public Uni<EmailInfo> takeFirstQueuedToProcessing() {
        return EmailInfo.find("status", EmailStatus.QUEUED)
                .firstResult()
                .onItem()
                .ifNotNull()
                .transformToUni(i -> {
                    EmailInfo x = (EmailInfo) i;
                    x.status = EmailStatus.PROCESSING;
                    return persist(x);
                });
    }
}
```

```Java
@Singleton
public class EmailService {

    @Inject
    EmailRepository repo;

    // ... Rest omitted

    // This shouldn't be an async API. It is blocking by definition, so instead use something like: processEmlFile
    public Uni<EmailDetails> archiveEmlFile(final String filePath, final EmailInfo emailInfo) {
        try {
            Properties props = new Properties();
            Session mailSession = Session.getInstance(props, null);
            FileInputStream stream = new FileInputStream(filePath);
            // BLOCKING (!) may take quite a bit to parse big emails - and yes there are such, with many MB of attachments etc.
            MimeMessage message = new MimeMessage(mailSession, stream);
            EmailDetails d = EmailDetails.from(message, emailInfo);
            return repo.persist(d);
        } catch (Exception ex) {
            return Uni.createFrom().failure(new EmailArchivingException(ex.getMessage()));
        }
    }

    // Better API, as this transports that it is indeed a blocking call. It is CPU-bound, not I/O bound, so async doesn't help us
    // It needs to be executed on a Workerpool-thread!
    public MimeMessage processEmlFile(final String filePath) {
        try {
            Properties props = new Properties();
            Session mailSession = Session.getInstance(props, null);
            FileInputStream stream = new FileInputStream(filePath);
            return new MimeMessage(mailSession, stream);
        } catch (Exception ex) {
            throw new EmailArchivingException(ex.getMessage());
        }
    }
}
```

```Java
@Entity
@Data
@Table(name = "email_info")
@Slf4j
@NoArgsConstructor
public class EmailInfo extends PanacheEntityBase {
    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "seq_email_info")
    @SequenceGenerator(name = "seq_email_info", sequenceName = "SEQ_EMAIL_INFO_ID", allocationSize = 1)
    @Column(name = "ID", nullable = false, unique = true)
    public Long id;

    @Column(name = "STORED_FILE_NAME", nullable = false)
    public String storedFilename;
    @Column(name = "FILE_HASH", nullable = false)
    public String fileHash;
    @Column(name = "FILE_SIZE", nullable = false)
    public BigInteger fileSize;

    @OneToOne(mappedBy = "emailInfo")
    @ToString.Exclude
    @JsonIgnore
    private EmailDetails emailDetails;
}
```

```Java
@ApplicationScoped
public class ShedlockRepository implements PanacheRepository<Shedlock> {
    public Uni<Shedlock> findById(String id) {
        return Shedlock.findById(id);
    }

    @WithTransaction
    public Uni<Boolean> delete(String id) {
        return Shedlock.deleteById(id);
    }

    @WithTransaction
    public Uni<Shedlock> persist(Shedlock shedlock) {
        return shedlock.persistAndFlush();
    }
}
```

```Java
@Entity
@Table(name = "SHEDLOCK")
@Data
@AllArgsConstructor
@NoArgsConstructor
public class Shedlock extends PanacheEntityBase {
    @Id
    @Column(name = "id")
    public String id;
    @Column(name = "lock_until")
    public LocalDateTime lockedUntil;
    @Column(name = "locked_at")
    public LocalDateTime lockedAt;
    @Column(name = "locked_by")
    public String lockedBy;
}
```

```Java
@ApplicationScoped
public class SchedulerCreator {

    @Inject
    org.quartz.Scheduler quartz;

    @Inject
    MailboxesService mailboxesService;

    @ConfigProperty(name = "cron.process-queued-emails")
    String cronProcessQueuedEmails;

 	void onStart(@Observes StartupEvent event) throws SchedulerException {
 		// ... create other jobs - code omitted
        var identity = "processQueuedEmails";

        JobDetail job = JobBuilder.newJob(ProcessQueuedEmailJob.class)
                .withIdentity(identity)
                .build();

        Trigger trigger = TriggerBuilder.newTrigger()
                .withIdentity(identity)
                .withSchedule(cronSchedule(cronProcessQueuedEmails))
                .startNow()
                .build();
        quartz.scheduleJob(job, trigger);
    }
}```

**Reply by Froidoh** [reply]
> @Froidoh not related to the core issue but I noticed that you are using the quartz extension and I am wondering if you've any requirement to use quartz extension? It seems that the requirement here is to create a clustered schedulers. The quartz extension has this but that's only work with a blocking data source and not a reactive one, so you cannot use that in the same app that's using the reactive data source (combining reactive and blocking data sources in the same app doesn't work). In my opinion, you could use the lightweight [scheduler extension](https://quarkus.io/guides/scheduler) for what you are looking to do with this scheduling need.

Hey, can you take a look at the (relevant) example (parts) I posted a minute ago?

We have no real requirement to do so. It's the thing the team is most comfortable with, as it's known to a few devs from legacy projects.

As you can see, we tried our best to create our own solution via a simple Entity and Repository.

We are using Quartz for quite a few cronjobs now.

The main reason to use Quartz was that the developer initially doing this had the requirement that the schedule needs to be configurable. As the docs didn't mention this on the "quarkus schedule" page, they decided to use the extension and manually schedule with a cron-expression taken from a config.

**Reply by machi1990** [reply]
Thanks @Froidoh we appreciate the details and the context.

There are suggestions from https://github.com/quarkusio/quarkus/discussions/33340#discussioncomment-5902207 and https://github.com/quarkusio/quarkus/discussions/33340#discussioncomment-5902219 can you try one of the two approaches?

**Reply by machi1990** [reply]
> The main reason to use Quartz was that the developer initially doing this had the requirement that the schedule needs to be configurable. As the docs didn't mention this on the "quarkus schedule" page, they decided to use the extension and manually schedule with a cron-expression taken from a config.

Thank you. As of Quarkus 3, it is possible to schedule and configure jobs programmatic with the lightweight scheduler extension too: https://quarkus.io/guides/scheduler-reference#programmatic_scheduling hence my suggestion

**Reply by Froidoh** [reply]
> Thanks @Froidoh we appreciate the details and the context.
> 
> There are suggestions from [#33340 (reply in thread)](https://github.com/quarkusio/quarkus/discussions/33340#discussioncomment-5902207) and [#33340 (reply in thread)](https://github.com/quarkusio/quarkus/discussions/33340#discussioncomment-5902219) can you try one of the two approaches?

I wanna say thanks @machi1990 and the others (don't want to ping all of you!) Thanks a bunch for helping out, for creating such a pleasant framework to work with, for being so responsive and for contributing so much to open source!

I am having a look at the suggested approaches. First step was to migrate to quarkus 3, to make life easier :) 

Next step is to find out how to combine all the new knowledge. I still think the easiest approach would probably be to just create a new microservice for the background tasks so we can be blocking and use hibernate-orm without being reactive at all, but I will have a hard time justifying this wrt management because a java pod, even with quarkus, still needs more resources than a node.js one (this is not a diss, that's the metric that is ruling at the moment -.-)

**Reply by mkouba** [reply]
@Froidoh note that the code from the `ProcessQueuedEmailJob#execute()` will not work in Quarkus 3 because a HR Panache entity method execution is not offloaded on the current Vert.x context anymore. Details are described in the [migration guide](https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.0#hibernate-reactive-panache).

You need to use either `Scheduler.JobDefinition.setAsyncTask()` or [`VertxContextSupport.subscribeAndAwait()`](https://quarkus.io/guides/vertx#executing-asynchronous-code-from-a-blocking-thread) from the blocking thread.

**Reply by Froidoh** [reply]
@mkouba 

Thank you

I am pretty sure I need to rewrite this in general - it is blocking by definition... at least the most important step in the middle is.
It's just the DB calls that could be async/non-blocking, but the cpu-bound task doesn't profit from being run on an eventloop... on the contrary, it is the main cause of headache.

Switching between eventloop and background threads - even if easily possible - seems like a waste here. The calls to the DB should be fast, they are not the performance bottleneck.

So to make life easier, to make the code easier, the approach I would take would be to run everything on the WorkerPool. As we established, that's not possible with hibernate-reactive and currently one cannot combine hibernate-reactive and hibernate-orm (blocking) in one quarkus application.

That is my dilemma right now.

I am sure we can come up with some smart workarounds, some code that switched the context in a safe manner... the question then is: if next week there is a new requirement, who can code it? Me? Maybe! Some one else? Not so much, as team-members are already complaining about the "non java"-syntax -.-

**Comment by Froidoh** [comment]
Currently investigating whether it's smarter (or even possible) to combine hibernate reactive and "hibernate blocking" in one and the same app.

Would mean that we have to double a lot of code to have a sync, blocking and an async, non blocking implementation (basically the coloring problem) but if it works... Right now because of the background jobs the pods die like flies.

The API endpoints that can be totally reactive and non-blocking work like a charm and we are happy with them (unless the background job blocks the eventloop again and all goes down)

**Comment by geoand** [comment]
Using blocking Hibernate for the background job would certainly make more sense. However at the time being I don't think we allow mixing Hibernate ORM and Hibernate Reactive in the same application (this restriction will likely be lifted in the future).

**Reply by Froidoh** [reply]
A bummer, maybe I need to extract this to another microservice for the time being then. Let's see. I'll await (pun intended) your judgement. We are obviously doing something very, very wrong here!
